<a href="https://colab.research.google.com/github/JulverRMP/Sales-Predictions-Food/blob/main/Proyecto_1_Parte_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

El primer proyecto será una predicción de ventas para productos alimenticios que se venden en distintas tiendas. El objetivo es ayudar a la tienda a comprender las propiedades de los productos y establecimientos que tienen un rol crucial en aumentar las ventas.

Para la parte 2, tu tarea es importar los datos y usar Pandas para comenzar a limpiar y explorar los datos. En la ciencia de datos, pasaremos cerca del 80% de nuestro tiempo preparando los datos(https://www.forbes.com/sites/gilpress/2016/03/23/data-preparation-most-time-consuming-least-enjoyable-data-science-task-survey-says/?sh=ba4f7346f637), ¡así que este es un paso muy importante! Primero, clona el repositorio GitHub que creaste para alojar tu proyecto, para que puedas asignarle este repositorio a tu trabajo. Ve el video al final de esta tarea si no tienes seguridad de cómo hacerlo.

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import pandas as pd

In [4]:
# Tenen cuenta que es probable que el nombre del archivo sea diferente en tu computadora (especialmente más adelante en el curso)
filename = '/content/drive/MyDrive/BOOTCAMP-DATA ANALYTICS/SEMANA02/sales_predictions.csv'
df = pd.read_csv(filename)

In [5]:
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

* Explora los datos: ¿Qué necesitas para limpiar estos datos?





verificar los tipos de datos, si hay datos nulos, datos erroneos, datos repetidos, realizar iputaciones de acuerdo a la verificación previo.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Según el método .info, se obtiene que las columnas donde tienen datos faltantes son: 'Item_Weight' , 'Outlet_Size'

In [7]:
df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

¿Hay valores faltantes en el conjunto de datos? Es posible que algunas tiendas no informen de todos los datos a causa de problemas técnicos u otros problemas. De ser así, trata con ello apropiadamente.

Verificando la informmación se observa que faltan datos en la columna 'Outlet_Size', por lo cual se identifica a que tiendas les falta dicho dato.

In [8]:
size_missing = df['Outlet_Size'].isna()
# Mira las filas que contienen NaN para size_missing
df.loc[size_missing,'Outlet_Identifier'].value_counts()

OUT045    929
OUT017    926
OUT010    555
Name: Outlet_Identifier, dtype: int64

Luego se observa que existe una relación entre el tipo de centro comercial, el tipo de area donde se ubica la tienda y el identificador de tienda y además las tiendas que no tienen una clasificación de tamaño son: OUT045, OUT017 y OUT010.

In [9]:
df.groupby(['Outlet_Type', 'Outlet_Location_Type', 'Outlet_Identifier'])[['Item_Identifier']].count()

Item_Identifier
Outlet_Type       Outlet_Location_Type Outlet_Identifier                 
Grocery Store     Tier 1               OUT019                         528
                  Tier 3               OUT010                         555
Supermarket Type1 Tier 1               OUT046                         930
                                       OUT049                         930
                  Tier 2               OUT017                         926
                                       OUT035                         930
                                       OUT045                         929
                  Tier 3               OUT013                         932
Supermarket Type2 Tier 3               OUT018                         928
Supermarket Type3 Tier 3               OUT027                         935

Por lo cual se procede a realizar un agrupamiento  según todas las categorias anteriores , añadiendo la coluna 'Outlet_Size' y relizando la suma de todas las ventas de las tiendas.

In [10]:
df.groupby(['Outlet_Type', 'Outlet_Location_Type', 'Outlet_Identifier','Outlet_Size'])[['Item_Outlet_Sales']].sum()

Item_Outlet_Sales
Outlet_Type       Outlet_Location_Type Outlet_Identifier Outlet_Size                   
Grocery Store     Tier 1               OUT019            Small             1.796941e+05
Supermarket Type1 Tier 1               OUT046            Small             2.118395e+06
                                       OUT049            Medium            2.183970e+06
                  Tier 2               OUT035            Small             2.268123e+06
                  Tier 3               OUT013            High              2.142664e+06
Supermarket Type2 Tier 3               OUT018            Medium            1.851823e+06
Supermarket Type3 Tier 3               OUT027            Medium            3.453926e+06

En la tabla anterior no se muestran las tiendas que no tienen clasificación en la columna 'Outlet_Size', por lo que procede a realizar la misma agrupación pero sin considerar la columna 'Outlet_Size' para poder saber las ventas totales de cada tienda y comparar con la tabla anterior para asi poder clasificar las tiendas faltantes.

In [11]:
df.groupby(['Outlet_Type', 'Outlet_Location_Type', 'Outlet_Identifier'])[['Item_Outlet_Sales']].sum()

Item_Outlet_Sales
Outlet_Type       Outlet_Location_Type Outlet_Identifier                   
Grocery Store     Tier 1               OUT019                  1.796941e+05
                  Tier 3               OUT010                  1.883402e+05
Supermarket Type1 Tier 1               OUT046                  2.118395e+06
                                       OUT049                  2.183970e+06
                  Tier 2               OUT017                  2.167465e+06
                                       OUT035                  2.268123e+06
                                       OUT045                  2.036725e+06
                  Tier 3               OUT013                  2.142664e+06
Supermarket Type2 Tier 3               OUT018                  1.851823e+06
Supermarket Type3 Tier 3               OUT027                  3.453926e+06

Por lo cual se puede con cluir que las tiendas faltantes de clasificar tendrian la siguiente clasificación:

Esta tienda pertenece a la categoria "Grocery Store" y su venta es mayor a la tienda OUT019 que tiene una clasificación de small, por lo cual se clasificara con Medium.
* OUT010  : Medium

Las tiendas pertenecen a la categoria Supermarket Type1, tipo de área dnde se ubican (Tier 2) y además los valores de ventas son menores al de la tienda OUT035 el cual tiene una categoria de Small, por lo tanto lo cual las tiendas se clasificara con la categoria de Small.
* OUT045  :  Small

* OUT017  :  Small


In [12]:
out010_filter=df['Outlet_Identifier'] == 'OUT010'
df.loc[out010_filter, 'Outlet_Size']=df.loc[out010_filter, 'Outlet_Size'].fillna('Medium')

In [13]:
out017_filter=df['Outlet_Identifier'] == 'OUT017'
out045_filter=df['Outlet_Identifier'] == 'OUT045'
df.loc[(out017_filter | out045_filter), 'Outlet_Size']=df.loc[(out017_filter | out045_filter), 'Outlet_Size'].fillna('Small')

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                8523 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


S verifica las variables en la columna 'Item_Fat_Content'

In [15]:
df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

Se regulariza los valores a solo Low Fat y Regular

In [16]:
df["Item_Fat_Content"].replace("LF", "Low Fat", inplace=True)
df["Item_Fat_Content"].replace("low fat", "Low Fat", inplace=True)
df["Item_Fat_Content"].replace("reg", "Regular", inplace=True)
df["Item_Fat_Content"].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

Luego se agrupa la identificación de cada producto , la columna de pesos y se realia el conteo para ver cuantos producto hay que no sean repetidos


In [17]:
df.groupby(['Item_Identifier', 'Item_Weight'])[['Item_Weight']].count()

Item_Weight
Item_Identifier Item_Weight             
DRA12           11.600                 6
DRA24           19.350                 5
DRA59           8.270                  6
DRB01           7.390                  2
DRB13           6.115                  5
...                                  ...
NCZ30           6.590                  6
NCZ41           19.850                 5
NCZ42           10.500                 5
NCZ53           9.600                  4
NCZ54           14.650                 5

[1555 rows x 1 columns]

Segun la tabla anterior se obtienen 1555 datos, sin considerar los datos de producto unico, no se repiten en otra tienda, además no tienen ingresado el peso.

Luego se realiza un conteo de valores en la columna 'Item_Identifier' donde si se consideran los productos qu no tiene ingresado el peso en la tabla de datos.

In [18]:
df['Item_Identifier'].value_counts().reset_index()

index  Item_Identifier
0     FDW13               10
1     FDG33               10
2     NCY18                9
3     FDD38                9
4     DRE49                9
...     ...              ...
1554  FDY43                1
1555  FDQ60                1
1556  FDO33                1
1557  DRF48                1
1558  FDC23                1

[1559 rows x 2 columns]

Por lo tanto en la última tabla se obtienen 1559 filas, es decir 1559; pero en la tabla anterior donde se consideraban los productos con datos de peso ingresado se obtuvo 1555 productos; por lo tanto se puede llegar a la conclusión de que existen 4 productos que no fueron ingresados sus pesos en la columna 'Item_Weight' y que no existen los mismos productos en otras tiendas, es decir son productos unicos.

Se procede a hacer una lista de los productos en mención

In [19]:
lisn1=[]
for x in df['Item_Identifier'].unique():
  lisn1.append(x)
print(lisn1)

['FDA15', 'DRC01', 'FDN15', 'FDX07', 'NCD19', 'FDP36', 'FDO10', 'FDP10', 'FDH17', 'FDU28', 'FDY07', 'FDA03', 'FDX32', 'FDS46', 'FDF32', 'FDP49', 'NCB42', 'DRI11', 'FDU02', 'FDN22', 'FDW12', 'NCB30', 'FDC37', 'FDR28', 'NCD06', 'FDV10', 'DRJ59', 'FDE51', 'FDC14', 'FDV38', 'NCS17', 'FDP33', 'FDO23', 'DRH01', 'NCX29', 'FDV20', 'DRZ11', 'FDX10', 'FDB34', 'FDK43', 'FDA46', 'FDC02', 'FDL50', 'FDM39', 'NCP05', 'FDV49', 'FDL12', 'FDS02', 'NCL17', 'FDM40', 'FDR13', 'FDA43', 'NCP18', 'FDK21', 'NCX54', 'DRK35', 'FDY21', 'FDI26', 'FDM20', 'FDV27', 'FDF09', 'FDY40', 'FDY45', 'FDC46', 'FDH19', 'FDZ03', 'DRH37', 'NCI17', 'FDJ58', 'FDH35', 'FDG02', 'NCZ18', 'FDC29', 'FDQ10', 'FDN48', 'FDL04', 'FDV25', 'FDD58', 'FDN04', 'FDV45', 'NCL18', 'FDR12', 'FDG20', 'FDZ55', 'FDQ49', 'FDN33', 'FDN27', 'FDW20', 'DRG27', 'DRI25', 'FDA44', 'NCR17', 'FDU04', 'FDF41', 'FDB56', 'FDT28', 'FDD10', 'FDW57', 'DRB48', 'FDP09', 'FDH14', 'FDA47', 'FDG12', 'DRE60', 'DRK49', 'FDD03', 'FDS52', 'FDW39', 'FDX34', 'FDV11', 'FDD17', 

In [20]:
id_filter=df['Item_Identifier'] == 'FDO60'
df.loc[id_filter]['Item_Weight']

1653    20.0
2897    20.0
5175    20.0
6085    20.0
6266    20.0
6277    20.0
7534     NaN
Name: Item_Weight, dtype: float64

In [21]:
id_filter=df['Item_Identifier'] == 'FDO60'
df.loc[id_filter, 'Item_Weight']

1653    20.0
2897    20.0
5175    20.0
6085    20.0
6266    20.0
6277    20.0
7534     NaN
Name: Item_Weight, dtype: float64

In [22]:
df.loc[id_filter]['Item_Weight'].max()

20.0

Luego se ingresan los datos faltantes a los productos que cuentan con un peso determinado, el cual fue ingresado en algunas tiendas, pero en otras por lo cual se encuentra vacio en la columna 'Item_Weight'.

Se realiza un barrido a la lista de productos creada, luego se filtra cada producto en el dataframe, para luego localizazr las filas con el mismo producto y si cuenta con algunas filas NaN se ingresa el valor de la otra fila que si cuente el peso para el mismo producto; pero si el producto es unico lo deja en vacio.

In [23]:
for x in lisn1:
  id_filter=df['Item_Identifier'] == x
  m=df.loc[id_filter]['Item_Weight'].max()
  df.loc[id_filter,'Item_Weight']=df.loc[id_filter]['Item_Weight'].fillna(m)

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                8519 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                8523 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [25]:
df[df['Item_Weight'].isna()]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
927            FDN52          NaN          Regular         0.130933   
1922           FDK57          NaN          Low Fat         0.079904   
4187           FDE52          NaN          Regular         0.029742   
5022           FDQ60          NaN          Regular         0.191501   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
927   Frozen Foods   86.9198            OUT027                       1985   
1922   Snack Foods  120.0440            OUT027                       1985   
4187         Dairy   88.9514            OUT027                       1985   
5022  Baking Goods  121.2098            OUT019                       1985   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
927       Medium               Tier 3  Supermarket Type3          1569.9564  
1922      Medium               Tier 3  Supermarket Type3          4434.2280  
4187      Medium               Tier 3  Supermarket Type3          3453.5046  
5022       Small               Tier 1      Grocery Store           120.5098

Luego de rellenar los valores faltantes según el criterio mencionado anteriorente nos quedarian solo 4 filas con valores nulos ya que estos productos son unicos.

Se verifica si hay datos duplicados con el método duplicate

In [29]:
df.duplicated().sum()

0

In [33]:
tabla_nan=df[df['Item_Weight'].isna()]
tabla_nan

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
927            FDN52          NaN          Regular         0.130933   
1922           FDK57          NaN          Low Fat         0.079904   
4187           FDE52          NaN          Regular         0.029742   
5022           FDQ60          NaN          Regular         0.191501   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
927   Frozen Foods   86.9198            OUT027                       1985   
1922   Snack Foods  120.0440            OUT027                       1985   
4187         Dairy   88.9514            OUT027                       1985   
5022  Baking Goods  121.2098            OUT019                       1985   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
927       Medium               Tier 3  Supermarket Type3          1569.9564  
1922      Medium               Tier 3  Supermarket Type3          4434.2280  
4187      Medium               Tier 3  Supermarket Type3          3453.5046  
5022       Small               Tier 1      Grocery Store           120.5098

Según la tala anterior hay 4 productos que no ha sido reemplazado el valor NaN en la tabla, debido a que son productos unicos y no hay dato similar para ingresar, ademas dichos productos pertenecen a diferente tipo de artículo.

Por lo tanto se procede a rellenar los valores vacios con la media de los productos, clasificandolo según el tipo de artículo.

In [36]:
lisn2=list(tabla_nan['Item_Type'])
lisn2

['Frozen Foods', 'Snack Foods', 'Dairy', 'Baking Goods']

In [40]:
for x in lisn2:
  type_filter=df['Item_Type'] == x
  mean=df.loc[type_filter]['Item_Weight'].mean()
  df.loc[type_filter,'Item_Weight']=df.loc[type_filter]['Item_Weight'].fillna(mean)

Se verifica que no hay valores NaN en la ase de datos.

In [41]:
df[df['Item_Weight'].isna()]

Empty DataFrame
Columns: [Item_Identifier, Item_Weight, Item_Fat_Content, Item_Visibility, Item_Type, Item_MRP, Outlet_Identifier, Outlet_Establishment_Year, Outlet_Size, Outlet_Location_Type, Outlet_Type, Item_Outlet_Sales]
Index: []

¿Cuáles son algunos resúmenes estadísticos, información agregada, u otras tendencias que puedes observar en los datos usando Pandas?

In [27]:
df.describe().T

count         mean          std       min  \
Item_Weight                8519.0    12.875420     4.646098     4.555   
Item_Visibility            8523.0     0.066132     0.051598     0.000   
Item_MRP                   8523.0   140.992782    62.275067    31.290   
Outlet_Establishment_Year  8523.0  1997.831867     8.371760  1985.000   
Item_Outlet_Sales          8523.0  2181.288914  1706.499616    33.290   

                                   25%          50%          75%           max  
Item_Weight                   8.785000    12.650000    16.850000     21.350000  
Item_Visibility               0.026989     0.053931     0.094585      0.328391  
Item_MRP                     93.826500   143.012800   185.643700    266.888400  
Outlet_Establishment_Year  1987.000000  1999.000000  2004.000000   2009.000000  
Item_Outlet_Sales           834.247400  1794.331000  3101.296400  13086.964800